src/distribution_spatial.cpp

```diff
@@ -180,6 +182,71 @@ Position SphericalIndependent::sample(uint64_t* seed) const
   return {x, y, z};
 }
 
+//==============================================================================
+// MeshSpatial implementation
+//==============================================================================
+
+MeshSpatial::MeshSpatial(pugi::xml_node node)
+{
+  // No in-tet distributions implemented, could include distributions for the barycentric coords
+  // Read in unstructured mesh from mesh_id value
+  int32_t mesh_id = std::stoi(get_node_value(node, "mesh_id"));
+  // Get pointer to spatial distribution
+  mesh_idx_ = model::mesh_map.at(mesh_id);
+
+  // Check whether mesh pointer points to a mesh
+  mesh_ptr_ = dynamic_cast<Mesh*>(model::meshes[mesh_idx_].get());
+  if (!mesh_ptr_) {fatal_error("Mesh passed to spatial distribution is not a mesh object"); }
+
+  int32_t n_bins = this->n_sources();
+  std::vector<double> strengths(n_bins, 0.0);
+
+  mesh_CDF_.resize(n_bins+1);
+  mesh_CDF_[0] = {0.0};
+  total_strength_ = 0.0;
+
+  // Create cdfs for sampling for an element over a mesh
+  // Volume scheme is weighted by the volume of each tet
+  // File scheme is weighted by an array given in the xml file
+  mesh_strengths_ = std::vector<double>(n_bins, 1.0);
+  if (check_for_node(node, "strengths")) {
+      strengths = get_node_array<double>(node, "strengths");
+      if (strengths.size() != n_bins) {
+        fatal_error(
+          fmt::format("Number of entries in the source strengths array {} does "
+                      "not match the number of entities in mesh {} ({}).",
+            strengths.size(), mesh_id, n_bins));
+      }
+      mesh_strengths_ = std::move(strengths);
+  }
+
+  if (get_node_value_bool(node, "volume_normalized")) {
+    for (int i = 0; i < n_bins; i++) {
+      mesh_strengths_[i] *= mesh_ptr_->volume(i);
+    }
+  }
+
+  total_strength_ = std::accumulate(mesh_strengths_.begin(), mesh_strengths_.end(), 0.0);
+
+  for (int i = 0; i < n_bins; i++) {
+    mesh_CDF_[i+1] = mesh_CDF_[i] + mesh_strengths_[i]/total_strength_;
+  }
+
+  if (fabs(mesh_CDF_.back() - 1.0) > FP_COINCIDENT) {
+    fatal_error(fmt::format("Mesh sampling CDF is incorrectly formed. Final value is: {}", mesh_CDF_.back()));
+  }
+  mesh_CDF_.back() = 1.0;
+}
+
+Position MeshSpatial::sample(uint64_t* seed) const
+{
+  // Create random variable for sampling element from mesh
+  double eta = prn(seed);
+  // Sample over the CDF defined in initialization above
+  int32_t elem_idx = lower_bound_index(mesh_CDF_.begin(), mesh_CDF_.end(), eta);
+  return mesh_ptr_->sample(seed, elem_idx);
```

Comment:
Just to document my challenges with the code; if one needs to use the mesh index sampled here to dependently sample over energy/direction then they must either use a costly function to find the mesh element id from position when sampling energy or return a second variable (elem_idx most likely) here. Returning a second variable here does not fit the formatting of the IndependentSource class so a small refactor may be required in the future to move this code into a new "DependentSource" class.

Reply:
Spot on. I found that too and have implemented such a method. Thanks for noting this here!